cmake/app/boilerplate.cmake

```diff
@@ -219,8 +219,16 @@ if(NOT BOARD_DIR)
   message(FATAL_ERROR "Invalid usage")
 endif()
 
-get_filename_component(BOARD_ARCH_DIR ${BOARD_DIR}      DIRECTORY)
-get_filename_component(ARCH           ${BOARD_ARCH_DIR} NAME)
+# Determine arch by the directory name after 'boards'
+# We search until we find 'boards' so we can have any number
+get_filename_component(PARENT_DIR ${BOARD_DIR} PATH)
+get_filename_component(DIR_NAME  ${PARENT_DIR} NAME)
+while(NOT ("boards" STREQUAL ${DIR_NAME}))
+  set(ARCH ${DIR_NAME})
+  get_filename_component(GPARENT_DIR ${PARENT_DIR} PATH)
+  get_filename_component(DIR_NAME   ${GPARENT_DIR} NAME)
+  set(PARENT_DIR ${GPARENT_DIR})
+endwhile(NOT ("boards" STREQUAL ${DIR_NAME}))
```

Comment:
Using 'while' and saying "We search until we find 'boards' so we can have any number" implies that we support arbitrary nesting, but we don't, we support one level of nesting as this commit demonstrates:

https://github.com/zephyrproject-rtos/zephyr/pull/12377/commits/b34cf2baf5314da83047a003ba6f8d0102dca1bc

The code and comments should reflect this.

Reply:
I can change the commit message, but I want the code to stay the same, such that regardless of what struct of dir, we determine ARCH based on looking for 'boards'.